install-poetry.py

```diff
@@ -0,0 +1,808 @@
+        self._bin_dir = bin_dir()
+        self._cursor = Cursor()
+
+    def allows_prereleases(self) -> bool:
+        return self._preview
+
+    def run(self) -> int:
+        if self._git:
+            version = self._git
+        elif self._path:
+            version = self._path
+        else:
+            version, current_version = self.get_version()
+
+        if version is None:
+            return 0
+
+        self.display_pre_message()
+        self.ensure_directories()
+
+        try:
+            self.install(version)
+        except subprocess.CalledProcessError as e:
+            print(colorize("error", "An error has occured: {}".format(str(e))))
+            print(e.output.decode())
+
+            return e.returncode
+
+        self._write("")
+        self.display_post_message(version)
+
+        return 0
+
+    def install(self, version, upgrade=False):
+        """
+        Installs Poetry in $POETRY_HOME.
+        """
+        self._write(
+            "Installing {} ({})".format(
+                colorize("info", "Poetry"), colorize("info", version)
+            )
+        )
+
+        env_path = self.make_env(version)
+        self.install_poetry(version, env_path)
+        self.make_bin(version)
+
+        self._overwrite(
+            "Installing {} ({}): {}".format(
+                colorize("info", "Poetry"),
+                colorize("b", version),
+                colorize("success", "Done"),
+            )
+        )
+
+        self._data_dir.joinpath("VERSION").write_text(version)
+
+        return 0
+
+    def uninstall(self) -> int:
+        if not self._data_dir.exists():
+            self._write(
+                "{} is not currently installed.".format(colorize("info", "Poetry"))
+            )
+
+            return 1
+
+        version = None
+        if self._data_dir.joinpath("VERSION").exists():
+            version = self._data_dir.joinpath("VERSION").read_text().strip()
+
+        if version:
+            self._write(
+                "Removing {} ({})".format(
+                    colorize("info", "Poetry"), colorize("b", version)
+                )
+            )
+        else:
+            self._write("Removing {}".format(colorize("info", "Poetry")))
+
+        shutil.rmtree(str(self._data_dir))
+        for script in ["poetry", "poetry.bat"]:
+            if self._bin_dir.joinpath(script).exists():
+                self._bin_dir.joinpath(script).unlink()
+
+        return 0
+
+    def make_env(self, version: str) -> Path:
+        self._overwrite(
+            "Installing {} ({}): {}".format(
+                colorize("info", "Poetry"),
+                colorize("b", version),
+                colorize("comment", "Creating environment"),
+            )
+        )
+
+        env_path = self._data_dir.joinpath("venv")
+
+        with temporary_directory() as tmp_dir:
+            subprocess.call(
+                [sys.executable, "-m", "pip", "install", "virtualenv", "-t", tmp_dir],
```

Comment:
We should do a check for pip, as pip is not always present. If it's not present we have one of two options.

1. Automatically execute `python -m ensurepip`. 
2. Error out, instructing user to install pip and hint at 1, since they may want to use distro specific installation. 

Alternatively, `virtualenv` wheel should be usable without installation I think. That would make a cleaner approach.

Reply:
https://github.com/python-poetry/poetry/issues/3843